overlord/snapshotstate/backend/backend.go

```diff
@@ -476,6 +455,52 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 	return nil
 }
 
+// pathsForSnapshot returns a list of absolute paths under 'snapDir' that should
+// be included in the snapshot (based on what directories exist).
+func pathsForSnapshot(snapDir string, snapshot *client.Snapshot) ([]string, error) {
+	dirExists := func(path string) (bool, error) {
+		exists, isDir, err := osutil.DirExists(path)
+		if err != nil {
+			return false, err
+		}
+
+		if exists && isDir {
+			return true, nil
+		}
+
+		if !exists {
+			logger.Debugf("Not saving %q in snapshot #%d of %q as it is does not exist.", path, snapshot.SetID, snapshot.Snap)
+		} else if !isDir {
+			logger.Noticef("Not saving %q in snapshot #%d of %q as it is not a directory.", path, snapshot.SetID, snapshot.Snap)
+		}
+
+		return false, nil
+	}
+
+	if exists, err := dirExists(snapDir); err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, nil
+	}
+
+	var snapshotPaths []string
+	revDir := filepath.Join(snapDir, snapshot.Revision.String())
+	if ok, err := dirExists(revDir); err != nil {
+		return nil, err
+	} else if ok {
+		snapshotPaths = append(snapshotPaths, revDir)
+	}
+
+	commonDir := filepath.Join(snapDir, "common")
+	if ok, err := dirExists(commonDir); err != nil {
+		return nil, err
+	} else if ok {
+		snapshotPaths = append(snapshotPaths, commonDir)
+	}
```

Comment:
minor: you could remove some code duplication by having a loop on `for _, subDir := range []string{commonDir, snapshot.Revision.String()} { ... }`